src/js/export.js

```diff
@@ -0,0 +1,238 @@
+  ];
+
+  function exportCollectionToFile(collID, colName){
+    cID = collID;
+    cName = colName;
+    var modal = createExportModal();
+
+    // Comma should probably not be the default delimiter
+    // since a lot of the entries contain commas.
+    selectDelimiterCSV.selectedIndex = 1;
+    selectDelimiterLeaf.selectedIndex = 3;
+
+    closeBtn.addEventListener('click', function() {
+      document.body.removeChild(modal);
+    }, false);
+
+    cancelBtn.addEventListener('click', function() {
+      document.body.removeChild(modal);
+    }, false);
+
+    exportBtn.addEventListener('click', (evt) => {
+      clearComplaintsExport();
+      filename = document.getElementById("exportFilename").value;
+      var filenameValid = isFilenameValid();
+      var delimitersValid = areDelimitersValid();
+      if(!filenameValid || !delimitersValid){
+        $(exportBtn).parent().append(
+          el("div.complaint.export", {text:"Incorrect input"})
+        );
+        return;
+      }
+
+      getTaxonomyAndExport();
+    }, false);
+
+  }
+
+  function getTaxonomyAndExport(taxonomy){
+    window.api.ajax("GET", window.api.host + "/v1/collection/" + cID + "/graph").done(graph => {
+      var taxonomy = getTaxonomy(graph);
+
+      window.api.ajax("GET", window.api.host + "/v1/collection/" + cID + "/entries").done(entries => {
+        var numberOfEntries = entries.length;
+        var generalInformation = new Set();
+        var rows = [];
+        var firstRow = setHeaders(entries, generalInformation, taxonomy);
+        rows.push(firstRow);
+
+        entries.forEach(function(entry) {
+          addEntry(rows, entry, generalInformation, taxonomy, numberOfEntries);
+        })
+
+      })
+    })
+  }
+
+  function getTaxonomy(graph) {
+    var taxonomy = new Set();
+    var edges = graph.edges;
+    edges.forEach(edge => {
+      var type = edge.type;
+      taxonomy.add(type);
+    });
+    return taxonomy;
+  }
+
+  function setHeaders(entries, generalInformation, taxonomy){
+    var firstRow = "";
+    Object.keys(entries[0]).forEach(function(key,index) {
+      generalInformation.add(key);
+      if(index > 0)
+        firstRow += CSVDelimiter;
+      firstRow += key;
+    });
+
+    taxonomy.forEach(leaf => {
+      firstRow += CSVDelimiter + leaf;
+    });
+    firstRow += "\n";
+    return firstRow;
+  }
+
+  function addEntry(rows, entry, generalInformation, taxonomy, numberOfEntries){
+    window.api.ajax("GET", window.api.host + "/v1/entry/" + entry.id + "/taxonomy").done(entryTaxonomy => {
+      var curRow = calculateCurRow(entry, entryTaxonomy, generalInformation, taxonomy);
+      rows.push(curRow);
+
+      // Check if all entries have been added ( > because the first row in rows contains the headers).
+      if(rows.length > numberOfEntries){
+        var csvContent = "";
+        rows.forEach(row => {
+          csvContent += row;
+        });
+        exportToCSV(filename + ".csv", csvContent);
+      }
+    })
+  }
```

Comment:
Move the ajax query to `api.js` and return a (composable) promise. The output from one `then(..)` is fed into the next `then(..)` (it depends on the implementation, but that's how it usually works).

```js
function toCSV(...) {
    return api.v1.entry.taxonomy(entry.id).then(entryTaxonomy => {
        return calculateCurRow(entry, entryTaxonomy, generalInformation, taxonomy)
    })
}
```
If you use promises, then it's possible to do this in `getTaxonomyAndExport`:
```js
var entryRows = entries.map(entry => {
    return toCSV(entry, generalInformation, taxonomy, numberOfEntries).then(csvRow => rows.push(csvRow))
})
Promise.all(entryRows).then(() => {
    var csvContent = rows.join('')
    exportToCSV(`${filename}.csv`, csvContent)
})
```